src/middlewared/middlewared/plugins/disk_/identify.py

```diff
@@ -17,68 +19,65 @@ async def device_to_identifier(self, name, disks=None):
         elif disk_data and disk_data['serial']:
             return f'{{serial}}{disk_data["serial"]}'
 
-        await self.middleware.run_in_thread(geom.scan)
-        klass = geom.class_by_name('PART')
-        if klass:
-            for g in filter(lambda v: v.name == name, klass.geoms):
-                for p in g.providers:
-                    if p.config is None:
-                        continue
-                    if p.config['rawtype'] in await self.middleware.call('disk.get_valid_zfs_partition_type_uuids'):
-                        return f'{{uuid}}{p.config["rawuuid"]}'
-
-        g = geom.geom_by_name('LABEL', name)
-        if g:
-            return f'{{label}}{g.provider.name}'
-
-        g = geom.geom_by_name('DEV', name)
-        if g:
+        xml = await self.middleware.call('geom.get_xml')
+        if not xml:
+            return ''
+
+        found = xml.find(f'.//class[name="PART"]/geom/provider[name="{name}"')
+        if found:
+            rawtype = found.find('./config/rawtype')
+            if rawtype and rawtype.text in await self.middleware.call('disk.get_valid_zfs_partition_type_uuids'):
+                return f'{{uuid}}{found.find("./config/rawuuid").text}'
+
+        found = xml.find(f'.//class[name="LABEL"]/geom/provider[name="{name}"]')
+        if found:
+            return f'{{label}}{name}'
+
+        found = xml.find(f'.//class[name="DEV"]/geom/provider/[name="{name}"]')
+        if found:
             return f'{{devicename}}{name}'
 
         return ''
 
-    def identifier_to_device(self, ident, geom_scan=True, geom_xml=None):
+    def identifier_to_device(self, ident, geom_xml=None):
         if not ident:
-            return None
+            return
 
-        search = self.RE_IDENTIFIER.search(ident)
+        search = RE_IDENTIFIER.search(ident)
         if not search:
-            return None
-
-        if geom_scan:
-            geom.scan()
+            return
 
-        tp = search.group('type')
         # We need to escape single quotes to html entity
+        tp = search.group('type')
         value = search.group('value').replace("'", '%27')
 
         if tp == 'uuid':
-            _find = f'.//config[rawuuid = "{value}"]/../../name'
+            _find = f'.//config[rawuuid="{value}"]/../../name'
             if geom_xml:
                 search = geom_xml.find(_find)
             else:
-                search = geom.class_by_name('PART').xml.find(_find)
+                search = self.middleware.call_sync('geom.get_class_xml', 'PART').find(_find)
```

Comment:
So we are certain we will have xml here but that doesn't hold at L23 ? Why the inconsistency ?